debugger (#32)
sstsimulator/sst-docs: docs/guides/features/debug.mdx

3. [Watchpoints and tracing](#watchpoints-and-tracing)
4. [Defining custom commands](#custom-commands)
5. [Recording and replaying commands](#record-and-replay)
6. [History and command line editing](#history-and-command-line-editing)
7. [Designing for debug](#designing-for-debug)
8. [Command Reference](#command-reference)
9. [Known Issues](#known-issues)
10. [Planned improvements](#planned-improvements)

## Support status
Early support for debugging was introduced in SST 14.0 but was signficantly expanded in SST 15.1 to support a single threaded debug console. With SST 16, the debug console was extended to support multi-thread and multi-rank debug. 

This document will focus on the SST 16 debug console which is nearly identical to SST 15.1 with the exception of the following features provided by SST 16:
- New console commands to support navigating threads and ranks
- User control for large output streams using pagination

The debug console command line interface will be subject to changes in upcoming releases. The currently planned CLI changes are noted in the examples.

## Enabling the interactive console

> There are two primary methods to enable an interactive console from the command line: (1) specify the simulation time and (2) specify a signal handler

Reviewer: ```suggestion
There are two primary methods to enable an interactive console from the command line: (1) specify the simulation time and (2) specify a signal handler.
```

> To start an SST debug session using the simulation time use `--interactive-start=[TIME]` option. If `TIME` is omitted, then the interactive console will be active at 0ns. 

Reviewer: ```suggestion
To start an SST debug session using the simulation time, use the `--interactive-start=[TIME]` option. If `TIME` is omitted, then the interactive console will be activated when simulation begins at 0ns. 
```

The commands below are compatible starting with with SST 15.1 unless otherwise noted.

```sh title="Invoking the interactive console using simulation time"
 $ sst --interactive-start configuration.py
 $ sst --interactive-start=1us configuration.py
 $ sst --interactive-start=1us --replay-file=init.sstdbg configuration.py
```

To enable entering the debug console in response to a signal use `--sigusr1=sst.rt.interactive` or `--sigusr2=sst.rt.interactive`

```sh title="Invoking the interactive console using a signal handler"
 $ sst --sigusr1=sst.rt.interactive configuration.py
 $ sst --sigusr2=sst.rt.interactive configuration.py
```

When the SST process is running, enter the interactive console by sending the appropriate signal to its process ID.

```sh title="Example sending signal to SST from a bash shell"
 $ kill -s sigusr1 $pid
 $ kill -s sigusr2 $pid
```

> For more information on using signal handlers with SST see [Signal Handling](signal.md) 

Reviewer: ```suggestion
For more information on using signal handlers with SST see [Signal Handling](signal.md).
```

It is often convenient to save commands to a file and replay them upon entering the debug console. The command line option `--replay-file=FILE` is provided for this purpose.

```sh title="Replaying debug commands when entering the interactive console"
 $ sst --interactive-start=1us --replay-file=init.sstdbg configuration.py
```

## Debug session example

> The following example demonstrates debugging a simulation with 2 ranks with 2 threads per rank. This example is applicable to SST 16.0 or newer releases. On SST 15.1, the interaction console options will be ignored for simulations using more than 1 thread or rank. 

Reviewer: ```suggestion
The following example demonstrates debugging a simulation with 2 ranks with 2 threads per rank. This example is applicable to SST 16.0 or newer releases. On SST 15.1, the interactive console options will be ignored for simulations using more than 1 thread or rank. 
```

```sh title="Entering the debug console for a multi-rank / multi-threaded simulation"
 $ mpirun -np 2 sst -n 2 --interactive-start=1us configuration.py

    INTERACTIVE CONSOLE
    -- Rank:0/2 Thread:0/2 (Triggered)
    -- Component Summary
    cp_0_0/ (SST::GridNode::GridNode)
    cp_0_1/ (SST::GridNode::GridNode)
    cp_1_0/ (SST::GridNode::GridNode)
    cp_1_1/ (SST::GridNode::GridNode)

    -- Rank:0/2 Thread:1/2 (Triggered)
    -- Component Summary
    cp_2_0/ (SST::GridNode::GridNode)
    cp_2_1/ (SST::GridNode::GridNode)
    cp_3_0/ (SST::GridNode::GridNode)
    cp_3_1/ (SST::GridNode::GridNode)

    -- Rank:1/2 Thread:0/2 (Triggered)
    -- Component Summary
    cp_4_0/ (SST::GridNode::GridNode)
    cp_4_1/ (SST::GridNode::GridNode)
    cp_5_0/ (SST::GridNode::GridNode)
    cp_5_1/ (SST::GridNode::GridNode)

    -- Rank:1/2 Thread:1/2 (Triggered)
    -- Component Summary
    cp_6_0/ (SST::GridNode::GridNode)
    cp_6_1/ (SST::GridNode::GridNode)
    cp_7_0/ (SST::GridNode::GridNode)
    cp_7_1/ (SST::GridNode::GridNode)

    ---- Rank0:Thread0: Entering interactive mode at time 1000000
    Interactive start at 1000000
>
```

> The debugger uses a bash-like directory navigation features such as `pwd`, `cd`, and `ls`.

Reviewer: ```suggestion
The debugger uses bash-like directory navigation features such as `pwd`, `cd`, and `ls`.
```

```sh title="Basic navigation (SST 16.0)"
> pwd
    ()
> ls
    cp_0_0/ (SST::GridNode::GridNode)
    cp_0_1/ (SST::GridNode::GridNode)
    cp_1_0/ (SST::GridNode::GridNode)
    cp_1_1/ (SST::GridNode::GridNode)
> pwd
    ()
> ls
    cp_0_0/ (SST::GridNode::GridNode)
    cp_0_1/ (SST::GridNode::GridNode)
    cp_1_0/ (SST::GridNode::GridNode)
    cp_1_1/ (SST::GridNode::GridNode)
> cd cp_0_1
> cd state
> pwd
    cp_0_1/state (std::vector<unsigned int, std::allocator<unsigned int> >)
> 
```

:::info
On SST 16, the `cd` command can only navigate up or down one level. 

Planned navigation improvements include:

> - The top level directory will be references using a forward slash ('/')

Reviewer: ```suggestion
- The top level directory will be referenced using a forward slash ('/')
```

- Support use of full and relative paths to elements
:::

> Navigation between ranks and threads is supported using the 'info', 'thread', and 'rank' commands.

Reviewer: ```suggestion
Navigation between ranks and threads is supported using the `info`, `thread`, and `rank` commands.
```

```sh title="Navigating between ranks and threads"
> info all
    Rank 0/2 Thread 0/2 (Process 52704)
    Rank 0/2 Thread 1/2 (Process 52704)
    Rank 1/2 Thread 0/2 (Process 52705)
    Rank 1/2 Thread 1/2 (Process 52705)

> info current
    Rank 0/2 Thread 0/2 (Process 52704)

> rank 1
    ---- Rank1:Thread0: Entering interactive mode at time 1000000
    Interactive start at 1000000

> thread 1
    ---- Rank1:Thread1: Entering interactive mode at time 1000000
    Interactive start at 1000000
> ls
    cp_6_0/ (SST::GridNode::GridNode)
    cp_6_1/ (SST::GridNode::GridNode)
    cp_7_0/ (SST::GridNode::GridNode)
    cp_7_1/ (SST::GridNode::GridNode)
```

Data members of the objects can be viewed using the `print -rN` command, where `N` refers to the maximum number of levels in the object's hierarchy (defaults to 4). Members that are fundamental types can be modified using the `set` command.

```sh title="Printing and setting a variable"
> print dataMax
dataMax = 33554431 (unsigned long)
> set dataMax 64 
> print dataMax
    dataMax = 64 (unsigned long)
```

:::info
The `print` command will be changing to `print [-r N] [-v N] [-f <base>] [<obj>][[idx][idx]]`.

> This planned improvement supports data formatting,  verbosity controls, and the ability to specify multiple array indices

Reviewer: ```suggestion
This planned improvement supports data formatting,  verbosity controls, and the ability to specify multiple array indices.
```

    ---- Rank0:Thread0: Entering interactive mode at time 2000000
    Ran clock for 1000 sim cycles
> cd cp_0_0

> print curCycle
    curCycle = 86 (unsigned long)

> watch curCycle < 10
    Added watchpoint #0

> run
    INTERACTIVE CONSOLE
    -- Rank:0/1 Thread:0/2 (Triggered)
    -- Component Summary
    cp_0_0/ (SST::GridNode::GridNode)
    cp_0_1/ (SST::GridNode::GridNode)
    cp_1_0/ (SST::GridNode::GridNode)
    cp_1_1/ (SST::GridNode::GridNode)
    cp_2_0/ (SST::GridNode::GridNode)
    cp_2_1/ (SST::GridNode::GridNode)
    cp_3_0/ (SST::GridNode::GridNode)
    cp_3_1/ (SST::GridNode::GridNode)

    -- Rank:0/1 Thread:1/2 (Not Triggered)
    -- Component Summary
    cp_4_0/ (SST::GridNode::GridNode)
    cp_4_1/ (SST::GridNode::GridNode)
    cp_5_0/ (SST::GridNode::GridNode)
    cp_5_1/ (SST::GridNode::GridNode)
    cp_6_0/ (SST::GridNode::GridNode)
    cp_6_1/ (SST::GridNode::GridNode)
    cp_7_0/ (SST::GridNode::GridNode)
    cp_7_1/ (SST::GridNode::GridNode)

    ---- Rank0:Thread0: Entering interactive mode at time 3000000
    Last Trigger: WP0: BC : cp_0_0/curCycle ...

> print curCycle
    curCycle = 41 (unsigned long)
```

:::important

> In parallel simulations, the interactive console is invoked only at simulator synchronization points. Therefore, the value of the watched variable at the synchronization point may be different from the watchpoint trigger condition. To deal with this sampling issue, the debugger supports tracing variables associated with watchpoint triggers. See [Watchpoint and tracing](#watchpoints-and-tracing) for more details.

Reviewer: ```suggestion
In parallel simulations, the interactive console is invoked only at simulator synchronization points. Therefore, the value of the watched variable at the synchronization point may be different from the watchpoint trigger condition. To deal with this sampling issue, the debugger supports tracing variables associated with watchpoint triggers. See [Watchpoints and tracing](#watchpoints-and-tracing) for more details.
```

:::

Finally, to cleanly exit an interactive console session there are 2 options: (1) Use the `quit` command to exit the debugger but continue the simulation and (2) Use the `shutdown` command to exit the debugger and perform an immediate shutdown of the simulation.

```sh title="Exiting the interactive console using 'quit'"
> quit
Do you want to delete all watchpoints? [yes, no]
yes
Removing all watchpoints and exiting ObjectExplorer
Simulation is complete, simulated time: 10 us
```

```sh title="Exiting the interactive console using 'shutdown'"
> shutdown
Simulation shutdown
R0, T0: Exiting ObjectExplorer and shutting down simulation
Simulation is complete, simulated time: 0 s
```

:::info
Using `quit` will only clear watchpoints if selected by the user. Any other event (e.g. a prior `run \[TIME]`]) that is already scheduled will still execute normally.
:::

## Watchpoints and tracing

> Watchpoints provide the ability to define and detect events resulting in breaking into the interactive debug console at simulation synchronization points. A synchronization point is a point in time of the SST simulation where all threads and processes are at a stable state during which the main thread can safely perform information exchange across all partitions (like generation of a checkpoint). 

Reviewer: ```suggestion
Watchpoints provide the ability to define and detect events which will cause SST to break into the interactive debug console at the next simulation synchronization point. A synchronization point is a point in time of the SST simulation where all threads and processes are at a stable state during which the main thread can safely perform information exchange across all partitions (like generation of a checkpoint). 
```

> Watchpoints are defined using simple sequences of boolean operations. For SST 16, the operations are parsed from left to right with no support for conventional order of operations or grouping using parenthesis. Since multiple watchpoints can be defined, grouping of operations can be done by defining multiple watchpoints. During execution, the result will effectively be an 'OR'ing of all watchpoint conditions.

Reviewer: ```suggestion
Watchpoints are defined using simple sequences of boolean operations. For SST 16, the operations are parsed from left to right with no support for conventional order of operations or grouping using parentheses. Since multiple watchpoints can be defined, grouping of operations can be done by defining multiple watchpoints. During execution, the result will effectively be an 'OR'ing of all watchpoint conditions.
```

```sh title="Logging interactive console commands"
> logging
    sst console commands will be logged to sst-console.out
> rank 1
> thread 1
> cd cp_6_0
> print curCycle
    curCycle = 73 (unsigned long)
> shutdown
Simulation shutdown
    R0, T0: Exiting ObjectExplorer and shutting down simulation
    Simulation is complete, simulated time: 0 s

$ cat sst-console.out 
rank 1
thread 1
cd cp_6_0
print curCycle
shutdown
```

This file can be renamed and edited for replaying in another simulation. Any line starting with `#` will be treated as a comment.

```sh title="Logging interactive console commands"
$ cat sst-console.in
# disable user confirmation prompts
confirm false
# Enter rank 2 thread 1
rank 1
thread 1
cd cp_6_0
# Print values
print curCycle
print component_state_
# Run 10ns and print values again
run 10ns
print curCycle
print component_state_
# All done
shutdown
```

:::info
The `confirm false` command is provided to enable  replay files and user defined commands to execute without pausing for confirmation prompts.
:::

> The file can now be replayed by either using the SST command line option, '--replay-file' or using the interactive console's `replay` command.

Reviewer: ```suggestion
The file can now be replayed by either using the SST command line option, `--replay-file` or using the interactive console's `replay` command.
```

    ... lines skipped

    ---- Rank0:Thread0: Entering interactive mode at time 1000000
    Interactive start at 1000000
    > replay sst-console.in

    > # disable user confirmation prompts
    > confirm false
    > # Enter rank 2 thread 1
    > rank 1
    ---- Rank1:Thread0: Entering interactive mode at time 1000000
    Interactive start at 1000000
    > thread 1
    ---- Rank1:Thread1: Entering interactive mode at time 1000000
    Interactive start at 1000000
    > cd cp_6_0
    > # Print values
    > print curCycle
    curCycle = 73 (unsigned long)
    > print component_state_
    component_state_ = 3 (SST::BaseComponent::ComponentState)
    > # Run 10ns and print values again
    > run 10ns

    INTERACTIVE CONSOLE
    ... lines skipped

    ---- Rank1:Thread1: Entering interactive mode at time 2000000
    Ran clock for 10000 sim cycles
    > print curCycle
    curCycle = 39 (unsigned long)
    > print component_state_
    component_state_ = 3 (SST::BaseComponent::ComponentState)
    > # All done
    > shutdown
    Simulation shutdown
    R0, T0: Exiting ObjectExplorer and shutting down simulation
    Simulation is complete, simulated time: 0 s   
```

:::info

> Invoking the replay file from the SST command line will causes `replay` command to be invoked upon the first entry into to the debug console.

Reviewer: ```suggestion
Invoking the replay file from the SST command line will cause the `replay` command to be invoked upon the first entry into to the debug console.
```

:::

## Designing for debug

The SST debugger leverages the same serialization macros used for checkpointing to make most variables accessible in the interactive console. If a design supports checkpointing and restart then it should also support debugging.  However, a design does not need to be checkpointable to support debugging. Serialization macros can be added incrementally to provide visibility into variables of interest to the designer. See [checkpoint and restart](checkpoint.mdx) for details on using serialization macros.

There may be reasons to protect data members from writing or even visiblity in the debugger. To support this there are two mapping options provided with the `SST_SER` macro:
  * `SerOption::map_read_only`: Do not allow a user to modify this object in an interactive console
  * `SerOption::no_map`: Do not serialize this object when mapping

One important consideration for the current debugger is that watchpoint and trace variables must have persistent references. Using a data member that may be deleted will lead to unpredictable behavior including crashing the simulator. This includes any new/deleted objects as well as elements in resizable containers. 

> :::important

Reviewer: ```suggestion
:::warning
```
Suggest making this a 'warning' so the callout is more visible.

> Never attempt to use data members that may be deleted in watchpoints or traces

Reviewer: ```suggestion
Never attempt to use data members that may be deleted in watchpoints or traces.
```

| list | ls | \[-l] \[-ll] | list objects at the current level |

**State**

| Command | Alias | Options | Description | Notes |
|---------|-------|---------|-------------|-------|
| time | tm |  | print current simulation time in cycles |
| print | p | \[-r N] obj | print objects for N levels of hierarchy (4 default) |
| set | s | var value | set the value of a variable |

**Watch/Trace**

| Command | Alias | Options | Description | Notes |
|---------|-------|---------|-------------|-------|
| watch | w | trigger | add a watchpoint to the watchlist |
| trace | t | trigger : bufSize postDelay : v1 \[...vN] : action |
| watchList | wl | prints the current list of watchpoints |
| addTraceVar | add | index var1 \[...varN] | adds additional trace variables to existing watchpoint |
| printWatchPoint | prw | index | prints a watchpoint |
| printTrace | prt | index | prints trace buffer for a watchpoint |
| resetTrace | rst | index | reset trace buffer for a watchpoint |
| setHandler | shn | index t1 \[... tN] | set a watchpoint's handler location(s) for trigger check and sampling |
| unwatch | uw | \[index] | remove 1 or all watchpoints |

> Notes:

Reviewer: For consistency, either end all of the bullet points with a '.' or none of them.

Author: added periods to all the lines

* A `trigger` can be a comparison or sequence of comparisons combined with a logic operation.
* A `comparison` can be `var op val` which compares a variable to a given value.
* A `comparison` can also be `var changed` which checks when a value changes.
* An `op` can be `<`, `<=`,`>`,`>=`,`==`, or '!='
* A `logicOp` can be `&&` or `||`
* The `watch` command creates a default watchpoint that breaks into interactive debug when triggered.
* The `trace` command creates a watchpoint with a trace buffer to trace a set of variables and trigger an `action`
* Supported `actions` are: `interactive`, `printTrace`, `checkpoint`, `set var val`, `printStatus`, and `shutdown`
* `setHandler` may set the following: `bc`(before clock) `ac` (after clock) `be` (before event) `ae` (after event)
* `setHandler may also set `all` to select all handlers (default)

> * *Use the `--checkpoint-enable` SST command line option when using `checkpoint` actions*

Reviewer: Is it intentional to have this and the following in italics? I assume for higher visibility?

Author: It was intentional but after looking at it again decided against it.
